pkg/controller/compute/address.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/provider-gcp/apis/compute/v1beta1"
+	gcp "github.com/crossplane/provider-gcp/pkg/clients"
+)
+
+// Error strings.
+const (
+	errNotAddress = "managed resource is not an Address"
+)
+
+// SetupAddress adds a controller that reconciles Address managed resources.
+func SetupAddress(mgr ctrl.Manager, o controller.Options) error {
+	name := managed.ControllerName(v1beta1.AddressGroupKind)
+
+	r := managed.NewReconciler(mgr,
+		resource.ManagedKind(v1beta1.AddressGroupVersionKind),
+		managed.WithExternalConnecter(&addressConnector{kube: mgr.GetClient()}),
+		managed.WithConnectionPublishers(),
+		managed.WithPollInterval(o.PollInterval),
+		managed.WithLogger(o.Logger.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(o.ForControllerRuntime()).
+		For(&v1beta1.Address{}).
+		Complete(ratelimiter.NewReconciler(name, r, o.GlobalRateLimiter))
+}
+
+type addressConnector struct {
+	kube client.Client
+}
+
+func (c *addressConnector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	projectID, opts, err := gcp.GetAuthInfo(ctx, c.kube, mg)
+	if err != nil {
+		return nil, err
+	}
+	s, err := compute.NewService(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+	return &addressExternal{kube: c.kube, Service: s, projectID: projectID}, errors.Wrap(err, errNewClient)
+}
+
+type addressExternal struct {
+	kube      client.Client
+	projectID string
+	*compute.Service
+}
+
+func (e *addressExternal) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1beta1.Address)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotAddress)
+	}
+
+	region := gcp.StringValue(cr.Spec.ForProvider.Region)
+	observed, err := e.Addresses.Get(e.projectID, region, meta.GetExternalName(cr)).Context(ctx).Do()
+
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetAddress)
+	}
+
+	//  addresses are always "up to date" because they can't be updated. ¯\_(ツ)_/¯
+	eo := managed.ExternalObservation{ResourceExists: true, ResourceUpToDate: true}
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	address.LateInitializeSpec(&cr.Spec.ForProvider, *observed)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return eo, errors.Wrap(err, errManagedAddressUpdate)
+		}
+	}
+
+	cr.Status.AtProvider = address.GenerateAddressObservation(*observed)
+
+	switch cr.Status.AtProvider.Status {
+	case v1beta1.StatusReserving:
+		cr.SetConditions(xpv1.Creating())
+	case v1beta1.StatusInUse, v1beta1.StatusReserved:
+		cr.SetConditions(xpv1.Available())
+	}
+
+	return eo, errors.Wrap(err, errManagedAddressUpdate)
```

Comment:
I believe that you should recreate this error message, because its value is `"cannot update managed GlobalAddress resource"`, which would mislead users because its referencing `GlobalAddress` instead of `Address`.

Also, I would recreate all errors that are being used from `GlobalAddress`, because if the value is changed there from `Address` to `GlobalAddress`, it will be an error with misleading message

Reply:
Will do. Always a balance on whether or not to duplicate these type of constants.

Reply:
I renamed the constants in the `globaladdress.go` and added the original constants to `address.go`.